File: src/jobsApplied/storage.js
const JOB_APPLICATIONS_STORAGE_KEY = "rightforme-job-applications";
const JOB_APPLICATION_REQUIRED_FIELDS = ["id", "company", "roleTitle", "status"];
const JOB_APPLICATION_STATUSES = [
  "Found",
  "Reviewing",
  "Apply",
  "Maybe",
  "Skip",
  "Applied",
  "Interviewing",
  "Rejected",
  "Offer",
  "Closed",
];
const FIT_RECOMMENDATIONS = ["Apply", "Maybe", "Skip"];

function getJobApplications() {
  const saved = localStorage.getItem(JOB_APPLICATIONS_STORAGE_KEY);

  if (!saved) {
    return [];
  }

  try {
    const records = JSON.parse(saved);
    return Array.isArray(records) ? records : [];
  } catch {
    return [];
  }
}

function addJobApplication(record) {
  validateJobApplication(record);

  const records = getJobApplications();
  if (records.some((job) => job.id === record.id)) {
    throw new Error(`Job application already exists: ${record.id}`);
  }

  const timestampedRecord = {
    ...record,
    updatedAt: record.updatedAt || new Date().toISOString(),
  };
  const nextRecords = [...records, timestampedRecord];
  saveJobApplications(nextRecords);
  return nextRecords;
}

function updateJobApplication(jobId, updates) {
  const records = getJobApplications();
  const index = records.findIndex((job) => job.id === jobId);

  if (index === -1) {
    throw new Error(`Job application not found: ${jobId}`);
  }

  const updated = { ...records[index], ...updates, id: jobId, updatedAt: new Date().toISOString() };
  validateJobApplication(updated);
  records[index] = updated;
  saveJobApplications(records);
  return updated;
}

function saveJobApplications(records) {
  localStorage.setItem(JOB_APPLICATIONS_STORAGE_KEY, JSON.stringify(records, null, 2));
}

function replaceJobApplications(records) {
  records.forEach(validateJobApplication);
  saveJobApplications(records);
  return records;
}

function validateJobApplication(record) {
  const missing = JOB_APPLICATION_REQUIRED_FIELDS.filter((field) => !String(record[field] || "").trim());

  if (missing.length) {
    throw new Error(`Missing required field(s): ${missing.join(", ")}`);
  }

  if (!JOB_APPLICATION_STATUSES.includes(record.status)) {
    throw new Error(`Invalid job status: ${record.status}`);
  }

  validateFitReview(record);
}

function validateFitReview(record) {
  const score = record.fitAnalysis?.fitScore ?? record.fitScore;
  const recommendation = record.fitAnalysis?.recommendation ?? record.fitRecommendation;

  if (String(score || "").trim()) {
    const numericScore = Number(score);
    if (!Number.isFinite(numericScore) || numericScore < 0 || numericScore > 100) {
      throw new Error("Fit score must be a number from 0 to 100.");
    }
  }

  if (String(recommendation || "").trim() && !FIT_RECOMMENDATIONS.includes(recommendation)) {
    throw new Error(`Invalid fit recommendation: ${recommendation}`);
  }
}

window.RightForMeJobsAppliedStorage = {
  getJobApplications,
  addJobApplication,
  updateJobApplication,
  replaceJobApplications,
};

File: tools/jobs_applied/storage.py
from __future__ import annotations

import json
from pathlib import Path
from typing import Any


ROOT_DIR = Path(__file__).resolve().parents[2]
DEFAULT_STORAGE_PATH = ROOT_DIR / "data" / "job-applications.json"
REQUIRED_FIELDS = ["id", "company", "roleTitle", "status"]
VALID_STATUSES = {
    "Found",
    "Reviewing",
    "Apply",
    "Maybe",
    "Skip",
    "Applied",
    "Interviewing",
    "Rejected",
    "Offer",
    "Closed",
}
FIT_RECOMMENDATIONS = {"Apply", "Maybe", "Skip"}


class JobApplicationValidationError(ValueError):
    """Raised when a job application record is missing required data."""


class JobApplicationNotFoundError(LookupError):
    """Raised when a job application record cannot be found by id."""


def read_job_applications(path: Path = DEFAULT_STORAGE_PATH) -> list[dict[str, Any]]:
    if not path.exists():
        return []

    records = json.loads(path.read_text(encoding="utf-8"))
    if not isinstance(records, list):
        raise ValueError(f"Expected a list of job applications in {path}.")

    return records


def add_job_application(
    record: dict[str, Any],
    path: Path = DEFAULT_STORAGE_PATH,
) -> list[dict[str, Any]]:
    validate_job_application(record)
    records = read_job_applications(path)

    if any(existing.get("id") == record["id"] for existing in records):
        raise JobApplicationValidationError(f"Job application already exists: {record['id']}")

    records.append(record)
    write_job_applications(records, path)
    return records


def update_job_application(
    job_id: str,
    updates: dict[str, Any],
    path: Path = DEFAULT_STORAGE_PATH,
) -> dict[str, Any]:
    records = read_job_applications(path)

    for index, existing in enumerate(records):
        if existing.get("id") == job_id:
            updated = {**existing, **updates, "id": job_id}
            validate_job_application(updated)
            records[index] = updated
            write_job_applications(records, path)
            return updated

    raise JobApplicationNotFoundError(f"Job application not found: {job_id}")


def validate_job_application(record: dict[str, Any]) -> None:
    missing = [
        field
        for field in REQUIRED_FIELDS
        if not str(record.get(field, "")).strip()
    ]

    if missing:
        names = ", ".join(missing)
        raise JobApplicationValidationError(f"Missing required field(s): {names}")

    if record.get("status") not in VALID_STATUSES:
        raise JobApplicationValidationError(f"Invalid job status: {record.get('status')}")

    validate_fit_review(record)


def validate_fit_review(record: dict[str, Any]) -> None:
    fit_analysis = record.get("fitAnalysis") or {}
    score = fit_analysis.get("fitScore", record.get("fitScore"))
    recommendation = fit_analysis.get("recommendation", record.get("fitRecommendation"))

    if str(score or "").strip():
        try:
            numeric_score = float(score)
        except (TypeError, ValueError) as exc:
            raise JobApplicationValidationError("Fit score must be a number from 0 to 100.") from exc

        if numeric_score < 0 or numeric_score > 100:
            raise JobApplicationValidationError("Fit score must be a number from 0 to 100.")

    if str(recommendation or "").strip() and recommendation not in FIT_RECOMMENDATIONS:
        raise JobApplicationValidationError(f"Invalid fit recommendation: {recommendation}")


def write_job_applications(
    records: list[dict[str, Any]],
    path: Path = DEFAULT_STORAGE_PATH,
) -> None:
    path.parent.mkdir(parents=True, exist_ok=True)
    path.write_text(json.dumps(records, indent=2) + "\n", encoding="utf-8")
